flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/WindowedStream.java

```diff
@@ -459,7 +545,28 @@ public WindowedStream(KeyedStream<T, K> input,
 	 * @param resultType Type information for the result type of the window function
 	 * @return The data stream that is the result of applying the window function to the window.
 	 */
-	public <R> SingleOutputStreamOperator<R> apply(R initialValue, FoldFunction<T, R> foldFunction, WindowFunction<R, R, K, W> function, TypeInformation<R> resultType) {
+	public <R> SingleOutputStreamOperator<R> apply(R initialValue, FoldFunction<T, R> foldFunction,
```

Comment:
I think we can already implement https://issues.apache.org/jira/browse/FLINK-3869 (see my last comment there) for the `ProcessWindowFunction`. I.e. don't have `apply(ReduceFunction, WindowFunction)` but `reduce(ReduceFunction, ProcessWindowFunction)` (same for fold).

Also, `apply(R, FoldFunction, WindowFunction)` has a bug, in that it is too restrictive, see the Jira issue I linked.